src/track/track.cpp

```diff
@@ -697,41 +706,65 @@ int Track::getAnalyzerProgress() const {
     return load_atomic(m_analyzerProgress);
 }
 
-void Track::setCuePoint(double cue) {
+void Track::setCuePoint(double position, Cue::CueSource source) {
     QMutexLocker lock(&m_qMutex);
-    if (compareAndSet(&m_cuePoint, cue)) {
-        // Store the cue point in a load cue
-        CuePointer pLoadCue;
-        for (const CuePointer& pCue: m_cuePoints) {
-            if (pCue->getType() == Cue::LOAD) {
-                pLoadCue = pCue;
-                break;
+
+    if (updateCuePoint(position, source)) {
+        markDirtyAndUnlock(&lock);
+        emit(cuesUpdated());
+    }
+}
+
+bool Track::updateCuePoint(double position, Cue::CueSource source) {
+    // If source is not manual, snap cue point to nearest beat.
+    if (source != Cue::MANUAL && position != -1.0) {
+        if (m_pBeats) {
+            double closest_beat = m_pBeats->findClosestBeat(position);
+            if (closest_beat != -1.0) {
+                position = closest_beat;
             }
         }
-        if (cue > 0) {
-            if (!pLoadCue) {
-                pLoadCue = CuePointer(new Cue(m_id));
-                pLoadCue->setType(Cue::LOAD);
-                pLoadCue->setSource(Cue::MANUAL);
-                connect(pLoadCue.get(), SIGNAL(updated()),
-                        this, SLOT(slotCueUpdated()));
-                m_cuePoints.push_back(pLoadCue);
-            }
-            pLoadCue->setPosition(cue);
-        } else {
-            disconnect(pLoadCue.get(), 0, this, 0);
-            m_cuePoints.removeOne(pLoadCue);
+    }
+
+    if (!compareAndSet(&m_cuePoint, position) && !compareAndSet(&m_cueSource, source)) {
+        // Nothing changed.
+        return false;
+    }
+
+    // Store the cue point in a load cue
+    CuePointer pLoadCue = findCueByType(Cue::LOAD);
+    if (position != 0.0 && position != -1.0) {
+        if (!pLoadCue) {
+            pLoadCue = CuePointer(new Cue(m_id));
+            pLoadCue->setType(Cue::LOAD);
+            connect(pLoadCue.get(), SIGNAL(updated()),
+                    this, SLOT(slotCueUpdated()));
+            m_cuePoints.push_back(pLoadCue);
         }
-        markDirtyAndUnlock(&lock);
-        emit(cuesUpdated());
+        pLoadCue->setPosition(position);
```

Comment:
@uklotzde I like your idea a lot. I have started to implement it, but I have a dilemma: should I rework `Cue` class so it inherits `CuePosition` class? This way we can reduce duplication of `{get,set}{Position,Source}()` functions, and also make `Cue` easily (implicitly) convertible to `CuePosition` (but not vice-versa).